unix/w0vncserver/portals/RemoteDesktop.cxx

```diff
@@ -0,0 +1,319 @@
+  params = g_variant_new("(a{sv})", &optionsBuilder);
+
+  screenCast->call("CreateSession", params, requestHandleToken.c_str(),
+                   std::bind(&RemoteDesktop::handleCreateSession,
+                             this, std::placeholders::_1));
+}
+
+void RemoteDesktop::closeSession()
+{
+  if (session && sessionStarted)
+    session->call("Close", nullptr, nullptr, nullptr);
+}
+
+void RemoteDesktop::selectSources()
+{
+  GVariant* params;
+  GVariantBuilder optionsBuilder;
+  std::string requestHandleToken;
+
+  requestHandleToken = screenCast->newHandle();
+
+  g_variant_builder_init(&optionsBuilder, G_VARIANT_TYPE_VARDICT);
+  g_variant_builder_add(&optionsBuilder, "{sv}", "types",
+                        g_variant_new_uint32(SRC_MONITOR));
+  g_variant_builder_add(&optionsBuilder, "{sv}", "cursor_mode",
+                        g_variant_new_uint32(CUR_EMBEDDED));
+  g_variant_builder_add(&optionsBuilder, "{sv}", "multiple",
+                         g_variant_new_boolean(false));
+
+  g_variant_builder_add(&optionsBuilder, "{sv}", "handle_token",
+                        g_variant_new_string(requestHandleToken.c_str()));
+
+  params = g_variant_new("(oa{sv})", sessionHandle.c_str(),
+                         &optionsBuilder);
+
+  screenCast->call("SelectSources", params, requestHandleToken.c_str(),
+                   std::bind(&RemoteDesktop::handleSelectSources,
+                             this, std::placeholders::_1));
+}
+
+void RemoteDesktop::start()
+{
+  std::string requestHandleToken;
+  GVariantBuilder optionsBuilder;
+  GVariant *params;
+
+  requestHandleToken = screenCast->newHandle();
+
+  g_variant_builder_init(&optionsBuilder, G_VARIANT_TYPE_VARDICT);
+  g_variant_builder_add(&optionsBuilder, "{sv}", "handle_token",
+                        g_variant_new_string(requestHandleToken.c_str()));
+
+  params = g_variant_new("(osa{sv})", sessionHandle.c_str(), "",
+                         &optionsBuilder);
+
+  screenCast->call("Start", params, requestHandleToken.c_str(),
+                   std::bind(&RemoteDesktop::handleStart,
+                             this, std::placeholders::_1));
+}
+
+void RemoteDesktop::openPipewireRemote()
+{
+  GVariantBuilder optionsBuilder;
+
+  g_variant_builder_init(&optionsBuilder, G_VARIANT_TYPE_VARDICT);
+
+  g_dbus_proxy_call_with_unix_fd_list(
+    screenCast->getProxy(),
+    "OpenPipeWireRemote",
+    g_variant_new("(oa{sv})",
+    sessionHandle.c_str(),
+    &optionsBuilder),
+    G_DBUS_CALL_FLAGS_NONE,
+    3000,
+    nullptr,
+    nullptr,
+    [](GObject* proxy, GAsyncResult *res, void *userData) {
+      RemoteDesktop* self = (RemoteDesktop*)(userData);
+      self->handleOpenPipewireRemote(proxy, res);
+    },
+    this);
+}
+
+void RemoteDesktop::handleCreateSession(GVariant *parameters)
+{
+  GVariant* result;
+  uint32_t response;
+  GVariant* sessionHandleVariant;
+
+  assert(sessionHandle.empty());
+
+  g_variant_get(parameters, "(u@a{sv})", &response, &result);
```

Comment:
Doesn't this need some error handling? What if the variant has the wrong contents?